### PR TITLE
Gate Prometheus and Jaegar behind flags.

### PR DESCRIPTION
**Pull request by mwhittaker**
#18

### CODE REVIEW

File: internal/impl/kube.go [hunk 1]
```diff
@@ -165,19 +171,23 @@ func GenerateKubeDeployment(image string, dep *protos.Deployment, cfg *KubeConfi
...
 	}
-	generated = append(generated, content...)
 
-	// Generate the Prometheus deployment info.
-	content, err = generatePrometheusDeployment(maps.Values(replicaSets), dep)
-	if err != nil {
-		return fmt.Errorf("unable to create kube deployment for the Prometheus service: %w", err)
+	if opts.Prometheus {
+		// Generate the Prometheus deployment info.
+		content, err = generatePrometheusDeployment(maps.Values(replicaSets), dep)
```
Comment (giautm) [inline internal/impl/kube.go:185]: This function also generates the ConfigMap for Prometheus. Please move the deployment to a separated function, and only turn it off.

Comment (giautm) [reply internal/impl/kube.go:185]: The reason for keeping ConfigMap is that it contains the targets to the services endpoint. And we can't be known after applying the change to update the config outside of the manifest.

Comment (giautm) [reply internal/impl/kube.go:185]: make sense. :D

Comment (mwhittaker) [reply internal/impl/kube.go:185]: Ah, I see. Good point. I think if the user doesn't provide the `--prometheus` flag, then it is on them to set up Prometheus and make sure it works even when they change services. If the `--prometheus` flag is given, maybe there's something we can do to make that easier, but we can do that in a future PR.

File: cmd/weaver-kube/deploy.go [hunk 2]
```diff
@@ -37,12 +37,48 @@ const (
...
+Container Image Names:
+  "weaver kube deploy" builds and uploads a container image. You need to
+  specify the name of the container using the "image" field inside the "kube"
+  section of the config file. For example, consider the following config file:
+
+      [serviceweaver]
+      binary = "./foo"
+
+      [kube]
+      registry = "docker.io/my_docker_hub_username/foo"
```
Comment (spetrovic77) [inline cmd/weaver-kube/deploy.go:64]: docker.io is optional. Let's remove it.

File: internal/impl/kube.go [hunk 3]
```diff
@@ -123,17 +123,38 @@ type ListenerOptions struct {
 	Public bool
 }
 
-// kubeConfig stores the configuration information for one execution of a
+// KubeConfig stores the configuration information for one execution of a
 // Service Weaver application deployed using the Kube deployer.
 type KubeConfig struct {
+	// Image is the name of the container image that "weaver kube deploy"
+	// builds and uploads. For example, if Image is "docker.io/alanturing/foo",
```
Comment (spetrovic77) [inline internal/impl/kube.go:130]: ditto. delete `docker.io`.

**Review by spetrovic77 — APPROVED** [review]
Looks great.

**Comment by rgrandl** [comment]
I think we should think a little bit more on how to do this. 

It makes sense to disable both prometheus and jaeger for testing (if part of a continuous integration thing). In that case, we should have a single flag that disable observability. 

Otherwise, we need ways for the user to easily plug their custom prometheus and jaeger service, which should probably be done through configuration. Otherwise, if they disable jaeger/prometheus through flags, there is no easy way (or no option at all) for them to provide an alternative. 

Prometheus:
* the Prometheus service will pull the metrics from the /metrics path from all pods that satisfy a regex (more specific the app name); 
* if someone wants to use their own Prometheus, then they can probably modify the .yaml file and look at the definition of the config map to understand how to set their own Prometheus to scrape the app pods. 
* if we disable generating Prometheus yaml file at all, the user will not be able to see that, and will have no idea how to change it. 

Jaeger:
* this is push based; the babysitters will export traces to the Jaeger service
* even if we don't generate yaml info for Jaeger, there is no way for the babysitter right now to understand where to send the traces; maybe we will need a config option with the name of the Jaeger endpoint; maybe we can do something similar for Prometheus then too

**Comment by mwhittaker** [comment]
Changed the PR to a draft for now.

**Comment by rgrandl** [comment]
We might want to close this PR. In https://github.com/ServiceWeaver/weaver-kube/pull/44 we added support to enable integration with external monitoring services. Prometheus for metrics, Jaeger for traces, Loki for logs and Grafana for dashboard and other things.

**Pull request closed without merging**